crates/bevy_macro_utils/src/lib.rs

```diff
@@ -108,6 +108,48 @@ impl BevyManifest {
     }
 }
 
+/// Finds an identifier that will not conflict with the specified set of tokens.
+/// If the identifier is present in `haystack`, extra characters will be added
+/// to it until it no longer conflicts with anything.
+///
+/// Note that the returned identifier can still conflict in niche cases,
+/// such as if an identifier in `haystack` is hidden behind an un-expanded macro.
+pub fn ensure_no_collision(value: Ident, haystack: TokenStream) -> Ident {
+    // Collect all the identifiers in `haystack` into a set.
+    let idents = {
+        // List of token streams that will be visited in future loop iterations.
+        let mut unvisited = vec![haystack];
+        // Identifiers we have found while searching tokens.
+        let mut found = HashSet::new();
```

Comment:
This still requires us to rerun the TokenStream multiple times if `ensure_no_collision` called multiple times in the same macro. IMO we shouldn't promote the use of this in `bevy_macro_utils` until we retain this `HashSet` between calls to this function.

Reply:
Proc macros have so much overhead, I highly doubt that micro-optimizations like this are going to matter.